File: packages/context/src/walker.ts
import { readdirSync, readFileSync, statSync } from "node:fs";
import { join, relative } from "node:path";
import ignore, { type Ignore } from "ignore";

const ALWAYS_SKIP = new Set([
  "node_modules",
  ".git",
  "dist",
  "build",
  ".oni",
  "__pycache__",
  ".next",
  ".turbo",
  ".cache",
  "coverage",
  ".nyc_output",
  ".vscode",
  ".idea",
]);

const INDEXED_EXTENSIONS = new Set([
  ".ts",
  ".tsx",
  ".js",
  ".jsx",
  ".mjs",
  ".cjs",
  ".py",
  ".json",
  ".md",
  ".yaml",
  ".yml",
  ".toml",
  ".css",
  ".html",
  ".sql",
  ".sh",
  ".rs",
  ".go",
]);

function loadIgnoreFile(projectDir: string, filename: string): string | null {
  try {
    return readFileSync(join(projectDir, filename), "utf-8");
  } catch {
    return null;
  }
}

export function walkProject(projectDir: string): string[] {
  const ig: Ignore = ignore.default();

  const gitignore = loadIgnoreFile(projectDir, ".gitignore");
  if (gitignore) ig.add(gitignore);

  const oniignore = loadIgnoreFile(projectDir, ".oniignore");
  if (oniignore) ig.add(oniignore);

  const files: string[] = [];

  function walk(dir: string): void {
    let entries: string[];
    try {
      entries = readdirSync(dir);
    } catch {
      return;
    }

    for (const entry of entries) {
      if (entry.startsWith(".") && entry !== ".gitignore" && entry !== ".oniignore") {
        // Skip hidden files/dirs (except ignore files themselves)
        if (ALWAYS_SKIP.has(entry)) continue;
        // Still skip other dotfiles
        continue;
      }

      if (ALWAYS_SKIP.has(entry)) continue;

      const fullPath = join(dir, entry);
      const relPath = relative(projectDir, fullPath);

      // Check gitignore
      if (ig.ignores(relPath)) continue;

      let stat;
      try {
        stat = statSync(fullPath);
      } catch {
        continue;
      }

      if (stat.isDirectory()) {
        walk(fullPath);
      } else if (stat.isFile()) {
        const ext = "." + entry.split(".").pop();
        if (INDEXED_EXTENSIONS.has(ext)) {
          files.push(relPath);
        }
      }
    }
  }

  walk(projectDir);
  return files.sort();
}

export function detectLang(filePath: string): string {
  const ext = filePath.split(".").pop() ?? "";
  const langMap: Record<string, string> = {
    ts: "typescript",
    tsx: "typescript",
    js: "javascript",
    jsx: "javascript",
    mjs: "javascript",
    cjs: "javascript",
    py: "python",
    rs: "rust",
    go: "go",
    json: "json",
    md: "markdown",
    yaml: "yaml",
    yml: "yaml",
    toml: "toml",
    css: "css",
    html: "html",
    sql: "sql",
    sh: "shell",
  };
  return langMap[ext] ?? "unknown";
}

File: crates/oni-context/src/walker.rs
use ignore::WalkBuilder;
use std::path::{Path, PathBuf};

const MAX_FILE_SIZE: u64 = 512 * 1024; // 512KB

const ALWAYS_SKIP: &[&str] = &[
    "node_modules",
    ".git",
    "target",
    "dist",
    "build",
    "__pycache__",
    ".DS_Store",
    ".oni",
    ".next",
    ".turbo",
    ".cache",
    "coverage",
    ".nyc_output",
    ".vscode",
    ".idea",
];

const INDEXED_EXTENSIONS: &[&str] = &[
    "ts", "tsx", "js", "jsx", "mjs", "cjs", "py", "rs", "go", "java", "c", "cpp", "h", "hpp",
    "json", "md", "yaml", "yml", "toml", "css", "html", "sql", "sh", "rb", "cs", "swift", "kt",
    "scala", "php",
];

#[derive(Debug, Clone)]
pub struct DiscoveredFile {
    pub path: PathBuf,
    pub language: String,
}

pub fn walk_project(root: &Path) -> Vec<DiscoveredFile> {
    let mut builder = WalkBuilder::new(root);
    builder
        .hidden(true)
        .ignore(true)
        .git_ignore(true)
        .git_global(false)
        .git_exclude(false)
        .follow_links(false);

    // Load .oniignore if present
    let oniignore = root.join(".oniignore");
    if oniignore.exists() {
        builder.add_ignore(&oniignore);
    }

    let mut files = Vec::new();

    for entry in builder.build().flatten() {
        let path = entry.path();

        // Skip entries whose name is in the always-skip list
        if let Some(name) = path.file_name().and_then(|n| n.to_str()) {
            if ALWAYS_SKIP.contains(&name) {
                continue;
            }
        }

        // Only process regular files
        let Ok(metadata) = entry.metadata() else { continue };
        if !metadata.is_file() {
            continue;
        }

        // Skip large files
        if metadata.len() > MAX_FILE_SIZE {
            continue;
        }

        // Filter by extension
        let Some(ext) = path.extension().and_then(|e| e.to_str()) else { continue };
        if !INDEXED_EXTENSIONS.contains(&ext) {
            continue;
        }

        let language = detect_lang(ext);
        files.push(DiscoveredFile { path: path.to_path_buf(), language });
    }

    files.sort_by(|a, b| a.path.cmp(&b.path));
    files
}

pub fn detect_lang(ext: &str) -> String {
    match ext {
        "ts" | "tsx" => "typescript",
        "js" | "jsx" | "mjs" | "cjs" => "javascript",
        "py" => "python",
        "rs" => "rust",
        "go" => "go",
        "java" => "java",
        "c" | "h" => "c",
        "cpp" | "hpp" => "cpp",
        "rb" => "ruby",
        "cs" => "csharp",
        "swift" => "swift",
        "kt" => "kotlin",
        "scala" => "scala",
        "php" => "php",
        "sh" => "shell",
        "json" => "json",
        "md" => "markdown",
        "yaml" | "yml" => "yaml",
        "toml" => "toml",
        "css" => "css",
        "html" => "html",
        "sql" => "sql",
        _ => "unknown",
    }
    .to_string()
}
